crd-generator/api/src/main/java/io/fabric8/crd/generator/v1/decorator/EnsureSingleStorageVersionDecorator.java

```diff
@@ -19,46 +19,39 @@
 import io.fabric8.kubernetes.api.model.ObjectMeta;
 import io.fabric8.kubernetes.api.model.apiextensions.v1.CustomResourceDefinitionSpecFluent;
 import io.fabric8.kubernetes.api.model.apiextensions.v1.CustomResourceDefinitionVersion;
-import io.fabric8.kubernetes.api.model.apiextensions.v1.CustomResourceDefinitionVersionBuilder;
 
-import java.util.concurrent.atomic.AtomicReference;
-import java.util.function.Predicate;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
 
 public class EnsureSingleStorageVersionDecorator
     extends CustomResourceDefinitionDecorator<CustomResourceDefinitionSpecFluent<?>> {
 
-  private final AtomicReference<String> storageVersion = new AtomicReference<>();
-
   public EnsureSingleStorageVersionDecorator(String name) {
     super(name);
   }
 
   @Override
   public void andThenVisit(CustomResourceDefinitionSpecFluent<?> spec, ObjectMeta resourceMeta) {
-    Predicate<CustomResourceDefinitionVersionBuilder> hasStorageVersion = v -> v.getStorage() != null && v.getStorage();
+    List<CustomResourceDefinitionVersion> versions = spec.buildVersions();
 
-    if (spec.hasVersions() && !spec.hasMatchingVersion(hasStorageVersion)) {
-      spec.editFirstVersion().withStorage(true).endVersion();
-    }
+    List<String> storageVersions = versions.stream()
```

Comment:
I recall that, at some point, someone unrolled the `stream`s into plain loops because of [this feature](https://github.com/fabric8io/kubernetes-client/blob/main/doc/CRD-generator.md#generate-crds-in-parallel) and I believe that we are not testing it quite enough.

As ugly as it may sound 😞 , have you tried those changes with that flag enabled?

Reply:
Yes, that's why I added `ParallelCRDGeneratorExamplesTest`. It should work for parallel generation, too.

You are also right it's not the nicest way, but it's also hard to improve the situation here.
The core problem is not that we are working e.g. with streams or a predicate or a atomic reference. (That's just optimization of the error output).

The real problem is that the decorator sort algorithm is not working as expected on "multi level after/before" dependencies.

That's why I changed...

```java
@Override
public Class<? extends Decorator>[] after() {
  return new Class[] { AddCustomResourceDefinitionResourceDecorator.class,
      AddCustomResourceDefinitionVersionDecorator.class, SetStorageVersionDecorator.class };
}
```
<details><summary>Toggle me to see Decorator order...</summary>

```
[main] TRACE io.fabric8.crd.generator.v1.decorator.AddCustomResourceDefinitionResourceDecorator [apiGroup=sample.fabric8.io, kind=Multiple, name=multiples.sample.fabric8.io, plural=multiples, scope=Cluster]
[main] TRACE io.fabric8.crd.generator.v1.decorator.AddCustomResourceDefinitionResourceDecorator [apiGroup=sample.fabric8.io, kind=Multiple, name=multiples.sample.fabric8.io, plural=multiples, scope=Cluster]
[main] TRACE io.fabric8.crd.generator.v1.decorator.AddCustomResourceDefinitionVersionDecorator [name:multiples.sample.fabric8.io, version:v1]
[main] TRACE io.fabric8.crd.generator.v1.decorator.AddCustomResourceDefinitionVersionDecorator [name:multiples.sample.fabric8.io, version:v2]
[main] TRACE io.fabric8.crd.generator.v1.decorator.AddSchemaToCustomResourceDefinitionVersionDecorator [name:multiples.sample.fabric8.io, version:v1]
[main] TRACE io.fabric8.crd.generator.v1.decorator.AddSchemaToCustomResourceDefinitionVersionDecorator [name:multiples.sample.fabric8.io, version:v2]
[main] TRACE io.fabric8.crd.generator.v1.decorator.EnsureSingleStorageVersionDecorator [name:multiples.sample.fabric8.io]
[main] TRACE io.fabric8.crd.generator.v1.decorator.EnsureSingleStorageVersionDecorator [name:multiples.sample.fabric8.io]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SetDeprecatedVersionDecorator [name:multiples.sample.fabric8.io, version:v1, deprecated:false, deprecationWarning:null]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SetDeprecatedVersionDecorator [name:multiples.sample.fabric8.io, version:v2, deprecated:false, deprecationWarning:null]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SetServedVersionDecorator [name:multiples.sample.fabric8.io, version:v1, served:true]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SetServedVersionDecorator [name:multiples.sample.fabric8.io, version:v2, served:true]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SetStorageVersionDecorator [name:multiples.sample.fabric8.io, version:v1, storage:false]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SetStorageVersionDecorator [name:multiples.sample.fabric8.io, version:v2, storage:true]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SortCustomResourceDefinitionVersionDecorator [name:multiples.sample.fabric8.io]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SortCustomResourceDefinitionVersionDecorator [name:multiples.sample.fabric8.io]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SortPrinterColumnsDecorator [name:multiples.sample.fabric8.io, version:v2]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SortPrinterColumnsDecorator [name:multiples.sample.fabric8.io, version:v1]
```
</details>

to
```java
@Override
public Class<? extends Decorator>[] after() {
  return new Class[] {
      CustomResourceDefinitionVersionDecorator.class
  };
}
```
<details><summary>Toggle me to see Decorator order...</summary>

```
[main] TRACE io.fabric8.crd.generator.v1.decorator.AddCustomResourceDefinitionResourceDecorator [apiGroup=sample.fabric8.io, kind=Multiple, name=multiples.sample.fabric8.io, plural=multiples, scope=Cluster]
[main] TRACE io.fabric8.crd.generator.v1.decorator.AddCustomResourceDefinitionResourceDecorator [apiGroup=sample.fabric8.io, kind=Multiple, name=multiples.sample.fabric8.io, plural=multiples, scope=Cluster]
[main] TRACE io.fabric8.crd.generator.v1.decorator.AddCustomResourceDefinitionVersionDecorator [name:multiples.sample.fabric8.io, version:v1]
[main] TRACE io.fabric8.crd.generator.v1.decorator.AddCustomResourceDefinitionVersionDecorator [name:multiples.sample.fabric8.io, version:v2]
[main] TRACE io.fabric8.crd.generator.v1.decorator.AddSchemaToCustomResourceDefinitionVersionDecorator [name:multiples.sample.fabric8.io, version:v1]
[main] TRACE io.fabric8.crd.generator.v1.decorator.AddSchemaToCustomResourceDefinitionVersionDecorator [name:multiples.sample.fabric8.io, version:v2]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SetDeprecatedVersionDecorator [name:multiples.sample.fabric8.io, version:v1, deprecated:false, deprecationWarning:null]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SetDeprecatedVersionDecorator [name:multiples.sample.fabric8.io, version:v2, deprecated:false, deprecationWarning:null]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SetServedVersionDecorator [name:multiples.sample.fabric8.io, version:v1, served:true]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SetServedVersionDecorator [name:multiples.sample.fabric8.io, version:v2, served:true]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SetStorageVersionDecorator [name:multiples.sample.fabric8.io, version:v1, storage:false]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SetStorageVersionDecorator [name:multiples.sample.fabric8.io, version:v2, storage:true]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SortPrinterColumnsDecorator [name:multiples.sample.fabric8.io, version:v2]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SortPrinterColumnsDecorator [name:multiples.sample.fabric8.io, version:v1]
[main] TRACE io.fabric8.crd.generator.v1.decorator.EnsureSingleStorageVersionDecorator [name:multiples.sample.fabric8.io]
[main] TRACE io.fabric8.crd.generator.v1.decorator.EnsureSingleStorageVersionDecorator [name:multiples.sample.fabric8.io]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SortCustomResourceDefinitionVersionDecorator [name:multiples.sample.fabric8.io]
[main] TRACE io.fabric8.crd.generator.v1.decorator.SortCustomResourceDefinitionVersionDecorator [name:multiples.sample.fabric8.io]
```
</details>

The `EnsureSingleStorageVersionDecorator` extends `CustomResourceDefinitionDecorator` and because of that it is working on the `CustomResourceDefinition` "level". Meanwhile the `CustomResourceDefinitionVersionDecorator` and `SetStorageVersionDecorator` extend `CustomResourceDefinitionVersionDecorator`, which means that they are working on the CustomResourceDefinitionVersion "level". 

So far as I understand the sorting algorithm of the decorators mixing of multiple levels is not supported.

Maybe we should improve this sorting algorithm (or document the behaviour)?
Or am I wrong?

Reply:
> Yes, that's why I added `ParallelCRDGeneratorExamplesTest`. It should work for parallel generation, too.

That's nice, but I'm afraid that the execution of the annotation processors in the tests is not the same as when triggered using the "real" annotation processor.
Probably, the only way to fully test is to spin a new module like this:
https://github.com/fabric8io/kubernetes-client/tree/main/crd-generator/test
but enabling parallelism.

Thanks a lot for the rest of the analysis, I agree with this:

> Maybe we should improve this sorting algorithm (or document the behaviour)?

I see that someone had already been through this trouble: https://github.com/fabric8io/kubernetes-client/blob/0c80a66d10bf7096864246ce453413e0da4edbd2/crd-generator/api/src/main/java/io/fabric8/crd/generator/Resources.java#L84-L90

Hopefully, we can get another pair of eyes from @metacosm or @iocanel here.

Reply:
> can we implement this in a seperate PR

that's fine for me